[Feature] 태풍 씬 연습모드 구현

hyeokbini opened the pull request:
##  📍이슈 번호
#56 태풍 씬 구현

## ✍️ 주요 변경 사항
- 연습모드 / 태풍 씬 제작을 완료했습니다

##  🙏 코드 리뷰 참고 사항
- 변경사항이 꽤 많은데, 한번쯤 짚고 가셨으면 하는 부분은 IInteractable 인터페이스를 추가한 부분입니다. 상호작용한 오브젝트에 추가적인 상호작용이 불가능하도록 제어하는 기능이며, 오브젝트 인터랙션 스크립트에 상속하여 hasInteracted 인터페이스 프로퍼티를 작성하여 사용하시면 됩니다. 
- TyphoonScene2 씬에 들어가시면 테스트해보실 수 있습니다. 
- 퀄업할 부분이 꽤 많은데, 추후에 논의해보면 좋을 것 같습니다. 온전히 기능 구현과 기능 연결에만 집중한 PR이라고 봐주시면 좋을 것 같습니다.
- 코드상으로 궁금한 부분이 있다면 언제든지 질문해주세요!

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

* **신규 기능**
  * 다양한 상호작용 오브젝트(전기 스위치, 전화기, TV, 창문 등)와의 상호작용 기능이 추가되었습니다.
  * 상호작용 완료 여부를 확인할 수 있는 인터페이스가 도입되었습니다.
  * 테이프 획득 및 창문에 테이프를 사용하는 기능이 구현되었습니다.
  * 스테이지 진행 상황을 관리하고, 완료 시 페이드 효과와 함께 클리어 UI가 표시됩니다.
  * UI 텍스트 패널을 즉시 비활성화하는 기능이 추가되었습니다.
  * 플레이어가 테이프를 획득했을 때 상태가 반영되도록 개선되었습니다.
  * 상호작용 순서와 상태를 체크하는 로직과 관련 UI 활성화 흐름에 진단용 로그가 추가되었습니다.

* **버그 수정**
  * 상호작용 중복 방지를 위한 체크 및 UI 활성화 흐름이 개선되었습니다.

* **기타**
  * 관련 프리팹 및 메타데이터 파일이 추가되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

hyeokbini committed:
[Feature] TV 켜기 구현

hyeokbini committed:
[Feature] 테이프 집기 구현

+ 태풍 player 스크립트를 만들어서, grab 시에 이벤트 시스템을 통해 테이프를 잡은 bool 함수 제어

hyeokbini committed:
[Chore] 문 위치 수정, 콜라이더 수정

hyeokbini committed:
[Feature] 붙여져 있는 테이프 구현

interaction을 통해 active시켜야 함
+ 맵이 너무 좁아서 가구 재배치

hyeokbini committed:
[Feature] 태풍 씬 연습모드 완성

추후 퀄업해야 할 부분은 팀원 테스트 후 피드백 받아 진행

hyeokbini committed:
[Chore]

버튼 텍스트 변경

jeongha-daisy reviewed (approved):
수정된 코드, 새로 생긴 스크립트 확인 후 동작 확인했습니다 수고하셨습니다 👍
플레이하면서 느낀 건 말씀하신대로 논의하고 퀄업하면서 사용자 경험을 높이면 좋을 것 같네요!!

추가적으로 궁금한 건 interacionHandler에서 hitObject로 특정하였기 때문에 그 오브젝트의 public 변수인 hasInteracted만 참조하면 되는 건 아닌지, 인터페이스를 상속받도록 하는 구조가 필요한 이유가 궁금합니다!

hyeokbini commented:
> 수정된 코드, 새로 생긴 스크립트 확인 후 동작 확인했습니다 수고하셨습니다 👍
> 플레이하면서 느낀 건 말씀하신대로 논의하고 퀄업하면서 사용자 경험을 높이면 좋을 것 같네요!!
> 
> 추가적으로 궁금한 건 interacionHandler에서 hitObject로 특정하였기 때문에 그 오브젝트의 public 변수인 hasInteracted만 참조하면 되는 건 아닌지, 인터페이스를 상속받도록 하는 구조가 필요한 이유가 궁금합니다!

=====================================
@jeongha-daisy 
그렇게 된다면 예를 들어 TV와 핸드폰이 있다고 가정해 보겠습니다. 이 2개는 모두 GameObject입니다. 그리고 각 오브젝트에 사용되는 TVscript와 Phonescript가 있을 때, 해당 스크립트 내 변수를 참조하려면 TV에 상호작용했을 때는 getcomponent를 통해 TVscript를, 핸드폰에 상호작용했을 때는 Phonescript를 가져와야 합니다. 각 오브젝트에 사용하는 스크립트 클래스는 다르니까요.

위 상황이라면 모든 경우의 수에서 hitobject가 무엇인지(어떤 오브젝트인지) 판단한 후, 해당 오브젝트에 알맞는 스크립트를 가져오는 경우의 수를 하드코딩해야 하지만, 인터페이스 멤버 변수를 정의하고 기존 클래스에 상속시키면 상속된 클래스(스크립트)들은 인터페이스 프로퍼티를 getcomponent로 받아올 수 있습니다. 현재 스크립트에서는 IInteractable 인터페이스를 통해 hasinteracted를 가져오는 것처럼요.

이해되지 않는 부분이 있으시면 코멘트에 더 달아주세요!

jeongha-daisy commented:
> > 수정된 코드, 새로 생긴 스크립트 확인 후 동작 확인했습니다 수고하셨습니다 👍
> > 플레이하면서 느낀 건 말씀하신대로 논의하고 퀄업하면서 사용자 경험을 높이면 좋을 것 같네요!!
> > 추가적으로 궁금한 건 interacionHandler에서 hitObject로 특정하였기 때문에 그 오브젝트의 public 변수인 hasInteracted만 참조하면 되는 건 아닌지, 인터페이스를 상속받도록 하는 구조가 필요한 이유가 궁금합니다!
> 
> ===================================== @jeongha-daisy 그렇게 된다면 예를 들어 TV와 핸드폰이 있다고 가정해 보겠습니다. 이 2개는 모두 GameObject입니다. 그리고 각 오브젝트에 사용되는 TVscript와 Phonescript가 있을 때, 해당 스크립트 내 변수를 참조하려면 TV에 상호작용했을 때는 getcomponent를 통해 TVscript를, 핸드폰에 상호작용했을 때는 Phonescript를 가져와야 합니다. 각 오브젝트에 사용하는 스크립트 클래스는 다르니까요.
> 
> 위 상황이라면 모든 경우의 수에서 hitobject가 무엇인지(어떤 오브젝트인지) 판단한 후, 해당 오브젝트에 알맞는 스크립트를 가져오는 경우의 수를 하드코딩해야 하지만, 인터페이스 멤버 변수를 정의하고 기존 클래스에 상속시키면 상속된 클래스(스크립트)들은 인터페이스 멤버 변수를 getcomponent로 받아올 수 있습니다. 현재 스크립트에서는 IInteractable 인터페이스를 통해 hasinteracted를 가져오는 것처럼요.
> 
> 이해되지 않는 부분이 있으시면 코멘트에 더 달아주세요!

아 이해했습니다!! 특정 오브젝트를 지정했더라도 거기 달려있는 스크립트는 다르기 때문이군요. 감사합니다 👍

Rakhyunn reviewed (approved):
연습모드 텍스트 순서대로 하면 잘 진행되는 것을 확인했습니다. 고생하셨습니다!!
이럴 확률은 낮긴 하지만 만약 플레이어가 만약 순서에 맞지 않게 상호작용 하면 어색해보이는 것은 수정할 필요가 있어 보입니다. 예를 들어 TV를 처음에 켜라고 텍스트가 뜨지만 전기 차단기를 먼저 상호작용을 하면 '테이프로 집어 창문을 보강하세요'라고 텍스트가 나오는데 상호작용 순서도 제어할 필요가 있다고 생각합니다!

hyeokbini commented:
> 연습모드 텍스트 순서대로 하면 잘 진행되는 것을 확인했습니다. 고생하셨습니다!!
> 이럴 확률은 낮긴 하지만 만약 플레이어가 만약 순서에 맞지 않게 상호작용 하면 어색해보이는 것은 수정할 필요가 있어 보입니다. 예를 들어 TV를 처음에 켜라고 텍스트가 뜨지만 전기 차단기를 먼저 상호작용을 하면 '테이프로 집어 창문을 보강하세요'라고 텍스트가 나오는데 상호작용 순서도 제어할 필요가 있다고 생각합니다!

앗 깜빡했네요... 저도 그 부분 인지하고 있었는데 순서 제어 로직을 추가하고 PR 올린다는 걸 깜빡했습니다. 병합하기 전에 로직 추가 커밋해서 병합할게요!

Rakhyunn commented:
> > 연습모드 텍스트 순서대로 하면 잘 진행되는 것을 확인했습니다. 고생하셨습니다!!
> > 이럴 확률은 낮긴 하지만 만약 플레이어가 만약 순서에 맞지 않게 상호작용 하면 어색해보이는 것은 수정할 필요가 있어 보입니다. 예를 들어 TV를 처음에 켜라고 텍스트가 뜨지만 전기 차단기를 먼저 상호작용을 하면 '테이프로 집어 창문을 보강하세요'라고 텍스트가 나오는데 상호작용 순서도 제어할 필요가 있다고 생각합니다!
> 
> 앗 깜빡했네요... 저도 그 부분 인지하고 있었는데 순서 제어 로직을 추가하고 PR 올린다는 걸 깜빡했습니다. 병합하기 전에 로직 추가 커밋해서 병합할게요!
> 
> 

아하 그렇군요! 그럼 저는 그거 추가하시면 될 거 같습니다. 수고하셨습니다~

hyeokbini committed:
[Feature] 순서 제어 로직 구현

IInteractable 인터페이스에 interactindex 프로퍼티 추가
IManagerObjCount 인터페이스 추가(이름은 마땅히 생각이 안나서 막 지었습니다...)
해당 인터페이스는 각 씬 매니저 스크립트에 상속 후 Objcount => currentstepcount 프로퍼티로 받아와서 playercontroller.cs에서 사용합니다

hyeokbini committed:
Merge branch 'develop' into Feature/#56/hyeokbini
